core/arch/arm/plat-stm32mp1/plat_tzc400.c

```diff
@@ -33,91 +86,468 @@ static enum itr_return tzc_it_handler(struct itr_handler *handler __unused)
+		region_cfg.top = region_base + region_size - 1;
+		region_cfg.filters = filter_mask(tzc_dev->ddata->nb_filters);
+
+		prop = fdt_getprop(fdt, pnode, "st,protreg", &len);
+		if (!prop || len != (2 * sizeof(uint32_t)))
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		switch (fdt32_to_cpu(prop[0])) {
+		case DT_TZC_REGION_S_NONE:
+			region_cfg.sec_attr = TZC_REGION_S_NONE;
+			break;
+		case DT_TZC_REGION_S_RD:
+			region_cfg.sec_attr = TZC_REGION_S_RD;
+			break;
+		case DT_TZC_REGION_S_WR:
+			region_cfg.sec_attr = TZC_REGION_S_WR;
+			break;
+		case DT_TZC_REGION_S_RDWR:
+			region_cfg.sec_attr = TZC_REGION_S_RDWR;
+			break;
+		default:
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+		region_cfg.ns_device_access = fdt32_to_cpu(prop[1]);
+
+		DMSG("%#08"PRIxVA" - %#08"PRIxVA" : Sec access %i NS access %#"PRIx32,
+		     region_cfg.base, region_cfg.top, region_cfg.sec_attr,
+		     region_cfg.ns_device_access);
+
+		if (stm32mp_tzc_region_append(tzc_dev, &region_cfg))
+			panic("Error adding region");
+
+		if (exclude_region_from_nsec(&region_cfg))
+			panic("Not able to exclude region");
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Adds a TZC region entry for each non-secure memory area defined by
+ * nsec_region_list. The function releases resources used to build this
+ * non-secure region list.
+ */
+static void add_carved_out_nsec(struct tzc_device *tzc_dev)
+{
+	struct tzc_region_non_sec *region = NULL;
+	struct tzc_region_non_sec *region_safe = NULL;
+
+	SLIST_FOREACH_SAFE(region, &nsec_region_list, link, region_safe) {
+		DMSG("%#08"PRIxVA" - %#08"PRIxVA" : Sec access %i NS access %#"PRIx32,
+		     region->region.base, region->region.top,
+		     region->region.sec_attr,
+		     region->region.ns_device_access);
+
+		if (stm32mp_tzc_region_append(tzc_dev, &region->region))
+			panic("Error adding region");
+
+		SLIST_REMOVE(&nsec_region_list, region,
+			     tzc_region_non_sec, link);
+		free(region);
+	};
+}
+
+static TEE_Result stm32mp_tzc_parse_fdt(struct tzc_device *tzc_dev,
+					const void *fdt, int node)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
-	void *base = phys_to_virt(TZC_BASE, MEM_AREA_IO_SEC, 1);
-	unsigned int region_index = 1;
-	const uint64_t dram_start = DDR_BASE;
-	const uint64_t dram_end = dram_start + CFG_DRAM_SIZE;
-	const uint64_t tzdram_start = CFG_TZDRAM_START;
-	const uint64_t tzdram_size = CFG_TZDRAM_SIZE;
-	const uint64_t tzdram_end = tzdram_start + tzdram_size;
+	struct io_pa_va base = { };
+	size_t reg_size = 0;
+	int offs = 0;
 
-	assert(base);
+	base.pa = fdt_reg_base_address(fdt, node);
+	if (base.pa == DT_INFO_INVALID_REG)
+		return TEE_ERROR_BAD_PARAMETERS;
 
-	tzc_init((vaddr_t)base);
-	tzc_dump_state();
+	reg_size = fdt_reg_size(fdt, node);
+	if (reg_size == DT_INFO_INVALID_REG_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	tzc_dev->pdata.name = fdt_get_name(fdt, node, NULL);
+	tzc_dev->pdata.base = io_pa_or_va_secure(&base, reg_size);
+
+	res = interrupt_dt_get(fdt, node, &tzc_dev->pdata.itr_chip,
```

Comment:
FYI: I can be more efficient to get al deferrable resources first before digging into the DT for other information.
Here: first get the clock and interrupt resources, then get IOMEM base/size, name ...